x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/QueryableBuiltInRolesUpgradeIT.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.upgrades;
+
+import org.apache.http.client.methods.HttpGet;
+import org.elasticsearch.client.Request;
+import org.elasticsearch.client.Response;
+import org.elasticsearch.client.RestClient;
+import org.elasticsearch.common.util.Maps;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.test.rest.ObjectPath;
+import org.elasticsearch.xpack.core.security.authz.store.ReservedRolesStore;
+import org.junit.Before;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.toSet;
+import static org.elasticsearch.common.xcontent.support.XContentMapValues.extractValue;
+import static org.elasticsearch.xpack.core.security.test.TestRestrictedIndices.INTERNAL_SECURITY_MAIN_INDEX_7;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
+
+public class QueryableBuiltInRolesUpgradeIT extends AbstractUpgradeTestCase {
+
+    private static final String QUERYABLE_BUILT_IN_ROLES_NODE_FEATURE = "security.queryable_built_in_roles";
+    public static final String INDEX_METADATA_QUERYABLE_BUILT_IN_ROLES_DIGEST = "queryable_built_in_roles_digest";
+
+    @Before
+    public void initializeReservedRolesStore() {
+        new ReservedRolesStore();
+    }
+
+    /**
+     * Test upgrades from an older cluster versions that do not support queryable built-in roles feature.
+     */
+    public void testBuiltInRolesSyncedOnClusterUpgrade() throws Exception {
+        final int numberOfNodes = 3; // defined in build.gradle
+        waitForNodes(numberOfNodes);
+
+        final Set<TestNodeInfo> nodes = collectNodeInfos(adminClient());
```

Comment:
Couldn't we use `clusterHasFeature` -- not blocking since the current approach works but `clusterHasFeature` seems like a better fit.

Reply:
The `clusterHasFeature` is not quite fitting my use case right out of the box. I need to know which nodes (based on the version) in the cluster support which features. Knowing that the whole cluster (or just some of the nodes)  support the queryable feature, is not very helpful. I need a distinction between old and new nodes and their supported features.

I could make it work similarly to `ParameterizedFullClusterRestartTestCase` and keep a copy of old cluster's feature service and then combine decision making with a new feature services (mixed and upgraded), but that also felt too hacky as I simply need to know supported `nodes_features` and their versions.